### PR TITLE
feat(rust): Improve JSON output of ockam node show

### LINKED ISSUE #6142: Improve `ockam node show --output json` by including more information
# Current behavior

If you run `ockam node show --output json` it currently does not show much information. Here's a screenshot.

Image

# Desired behavior

It should display information that is available when running the command with plain text output `ockam node show`.

Please take a look at this [file](https://github.com/build-trust/ockam/blob/develop/implementations/rust/ockam/ockam_command/src/node/show.rs#L78) to get started.

---

<i>We love helping new contributors! ❤️
If you have questions or need help as you explore, please join us on [Discord](https://discord.gg/bewvnm6zqS). If you're looking for other issues to contribute to, please checkout our [good first issues](https://github.com/build-trust/ockam/labels/good%20first%20issue).</i>

### PR DESCRIPTION
**Pull request by kriogenia**
## Current behavior

Running `ockam node show --output json` currently only shows the name of the node instead of all the information shown with the plain option

## Proposed changes

Closes #6142

The PR creates a new struct in the `ockam_api` with the information to show in the `ockam node show` command. This new struct implements the `Output` trait with the plain printing logic. It implements Serialize to build the JSON response. Both response are now printed using the `OutputFormat`.

Added `Serialize` to all components that are being printed in the command, mimicking the same information as format as what was currently being printed via plain.  

Example of plain output and new JSON output:

```
Node:
  Name: start (default)
  Status: UP
  Route To Node:
    Short: /node/start
    Verbose: /dnsaddr/localhost/tcp/44041
  Identity: I1b3787e02c02453fe5cd1be07e1fd62b5ca466ff
  Transports:
    Transport:
      Type: TCP
      Mode: Listening
      Socket: 127.0.0.1:44041
      Worker: <none>
      FlowControlId: c0b280ee5872661f857639dd3b5bded7
  Secure Channel Listeners:
    Listener:
      Address: /service/api
      FlowControlId: 66ee5fe42fab11fb1b7121ab086e30d0
  Inlets:
  Outlets:
  Services:
    Service:
      Type: uppercase
      Address: /service/uppercase
    Service:
      Type: echo
      Address: /service/echo
    Service:
      Type: credentials
      Address: /service/credentials
 ```

```json
{
  "is_default": true,
  "name": "start",
  "is_up": true,
  "route": {
    "short": "/node/start",
    "verbose": "/dnsaddr/localhost/tcp/44041"
  },
  "identity": "I1b3787e02c02453fe5cd1be07e1fd62b5ca466ff",
  "transports": [
    {
      "type": "Tcp",
      "mode": "Listen",
      "socket": "127.0.0.1:44041",
      "worker": "<none>",
      "flow_control": {
        "id": "c0b280ee5872661f857639dd3b5bded7"
      }
    }
  ],
  "secure_channel_listeners": [
    {
      "address": "/service/api",
      "flow_control": {
        "id": "66ee5fe42fab11fb1b7121ab086e30d0"
      }
    }
  ],
  "inlets": [],
  "outlets": [],
  "services": [
    {
      "address": "/service/uppercase",
      "type": "uppercase"
    },
    {
      "address": "/service/echo",
      "type": "echo"
    },
    {
      "address": "/service/credentials",
      "type": "credentials"
    }
  ]
}
```

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] There are no Merge commits in this Pull Request. Ockam repo maintains a linear commit history. We merge Pull Requests by rebasing them onto the develop branch. Rebasing to the latest develop branch and force pushing to your Pull Request branch is okay.
- [x] I have read and accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have read and accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/.github/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam/blob/develop/.github/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam](https://github.com/build-trust/ockam) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam/edit/develop/.github/CONTRIBUTORS.csv) file in the github web UI and create a separate Pull Request, this will mark the commit as verified.

### CODE REVIEW

File: implementations/rust/ockam/ockam_command/tests/bats/nodes.bats [hunk 1]
```diff
@@ -41,9 +41,14 @@ force_kill_node() {
   run_success "$OCKAM" node create "$n"
 
   run_success "$OCKAM" node show "$n"
-  assert_output --partial "/dnsaddr/localhost/tcp/"
+  assert_output --partial "/dnsaddr/localhost/tcp/44041"
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/tests/bats/nodes.bats]: The assigned port is random. We just want to make sure it has an address, so it's OK to revert it back.

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs [hunk 2]
```diff
@@ -150,11 +150,16 @@ impl CreateOutlet {
...
     #[n(1)] pub bind_addr: String,
+    #[serde(skip)]
     #[n(2)] pub worker_addr: String,
+    #[serde(skip)]
     #[n(3)] pub alias: String,
     /// An optional status payload
+    #[serde(skip)]
     #[n(4)] pub payload: Option<String>,
+	#[serde(rename = "route_to_outlet", serialize_with = "super::address::serialize_route_as_multiaddr")]
     #[n(5)] pub outlet_route: String,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: There is a major problem with this approach here: the skip/rename rules you are adding here are tailored to match one specific use case: printing this struct in the CLI. So, instead of doing that, I'd propose to create a new struct in the `ockam_command` crate with just the fields that you need.

The same goes for the `OutletStatus` struct.

File: implementations/rust/ockam/ockam_api/src/nodes/models/address.rs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+{
+    s.serialize_some(&addr_to_multiaddr(t.clone()))
+}
+
+pub fn serialize_route_as_multiaddr<S>(route: &String, s: S) -> Result<S::Ok, S::Error>
+where
+S: Serializer,
+{
+	s.serialize_some(&Route::parse(route).and_then(|r| route_to_multiaddr(&r)))
+}
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/address.rs]: Not ideal, but I'd move this to `utils.rs` where we have the rest of multiaddr utility functions. We'll probably move all those functions to its own file, but we can do that in a separate PR.

File: implementations/rust/ockam/ockam_api/src/nodes/models/show.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
+use ockam_multiaddr::{
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/show.rs]: This whole file should be moved to `ockam_command`.

File: implementations/rust/ockam/ockam_api/src/nodes/models/show.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+#[derive(Debug, Serialize)]
+pub struct ShowNodeResponse<'a> {
+    pub is_default: bool,
+    pub name: &'a str,
+    pub is_up: bool,
+    pub route: RouteToNode,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub identity: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none", flatten)]
+    pub tcp_listeners: Option<TransportList>,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/show.rs]: Would it work if you add the `rename` attribute here insteaf of in the `TransportList`? If you can do that, you would be able to leave the `ockam_api` structs unmodified, which would be desirable.

File: implementations/rust/ockam/ockam_command/src/node/show.rs [hunk 6]
```diff
@@ -180,77 +63,51 @@ pub async fn print_query_status(
...
-        );
-    } else {
-        let node_state = cli_state.nodes.get(node_name)?;
-        // Get short id for the node
-        let default_id = match node_state.config().identity_config() {
-            Ok(resp) => resp.identifier().to_string(),
-            Err(_) => String::from("None"),
-        };
+    let node_state = cli_state.nodes.get(node_name)?;
+    let node_port = node_state
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/node/show.rs]: Note that the port won't be available until the node is up and running. Also note that the `is_node_up` function is looping a few times to wait unti the node is up, that's why this function must be the first to run before querying any node information. Same problem with the `node_state`: if the node is not up when you call it, you won't be getting all the node information.

File: implementations/rust/ockam/ockam_command/src/node/show.rs [hunk 7]
```diff
@@ -180,77 +63,51 @@ pub async fn print_query_status(
...
 
-    Ok(())
+            // Get list of outlets
+            let outlets: OutletList = node.ask(ctx, api::list_outlets()).await?;
+            node_info.outlets = Some(outlets);
+
+            node_info
+        };
+
+    Ok(opts.global_args.output_format.println_value(&node_info)?)
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/node/show.rs]: Since you now have an instance that implements the `Output` trait, the correct way to print the command output is throught the `terminal` instance. It would look something like:

```rust
opts.terminal
    .stdout()
    .plain(node_info)
    .json(serde_json::to_string_pretty(&node_info)?)
    .write_line()?;
```


[Here is an exampe](https://github.com/build-trust/ockam/blob/1a49eba84f17ea3b368aa22a1ffc6e53a474b474/implementations/rust/ockam/ockam_command/src/relay/delete.rs#L58) of another command using it.

**Review by adrianbenavides — COMMENTED** [review]
Hey @kriogenia , the PR looks really great, I was looking forward to this refactor! I've left some comments with some suggestions.

Also, don't forget to run `cargo fmt` and `cargo clippy`, it looks like there are some files without the empty line at the end.

File: implementations/rust/ockam/ockam_api/src/nodes/models/show.rs (hunk 5 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_api/src/nodes/models/show.rs]: That was my first attempt but flatten doesn't work with rename. It makes as the struct being flattened could hold multiple attributes instead of just one like we have. A possible workaround to leave the API structs unmodified would be using a custom serializer or taking out the `Vec` with the list instead of passing the whole `TransportList`, but I would say that in this particular cases specifying the name of the lists to be more meaningful like "transports" or "services" vs. just "list" is probably good.

File: implementations/rust/ockam/ockam_api/src/nodes/models/show.rs (hunk 4 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_api/src/nodes/models/show.rs]: Should I dump all its content (and the new serialization structs) to the `show.rs` file or do I create a new `model` module in `node` like is in `ockam_api` to store all this serialization logic?

File: implementations/rust/ockam/ockam_command/tests/bats/nodes.bats (hunk 1 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_command/tests/bats/nodes.bats]: Reverted

File: implementations/rust/ockam/ockam_api/src/nodes/models/address.rs (hunk 3 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_api/src/nodes/models/address.rs]: Moved

File: implementations/rust/ockam/ockam_api/src/nodes/models/show.rs (hunk 4 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_api/src/nodes/models/show.rs]: The current version has the models/show.rs version

File: implementations/rust/ockam/ockam_command/src/node/show.rs (hunk 6 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_command/src/node/show.rs]: Reverted to previous way of doing it

File: implementations/rust/ockam/ockam_command/src/node/show.rs (hunk 7 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_command/src/node/show.rs]: Fixed

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs [hunk 8]
```diff
@@ -150,11 +150,16 @@ impl CreateOutlet {
...
     #[n(1)] pub bind_addr: String,
+    #[serde(skip)]
     #[n(2)] pub worker_addr: String,
+    #[serde(skip)]
     #[n(3)] pub alias: String,
     /// An optional status payload
+    #[serde(skip)]
     #[n(4)] pub payload: Option<String>,
+	#[serde(rename = "route_to_outlet", serialize_with = "crate::util::serialize_route_as_multiaddr")]
     #[n(5)] pub outlet_route: String,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: We should revert the serde customizations for this struct and the OutletStatus. We use them in other crates "as is", so they rely on the current unmodified structure, and these changes would break that code.

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs [hunk 9]
```diff
@@ -237,10 +246,11 @@ impl OutletStatus {
...
 
 /// Response body when returning a list of Inlets
-#[derive(Debug, Clone, Decode, Encode)]
+#[derive(Debug, Clone, Decode, Encode, Serialize)]
 #[rustfmt::skip]
 #[cbor(map)]
 pub struct InletList {
+	#[serde(rename = "inlets")]
     #[n(1)] pub list: Vec<InletStatus>
 }
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: This still looks off to me... I think the proper way of doing this would be to delete all these `XList` structs that contain a single field with a vector, and refactor the endpoints that use them to return the plain vector. That way, you don't need to mess with serde anymore and you can customize the `ockam node show` structure as you want.

Since this is a little out of the scope for this PR, let me know if you want to tackle that as part of this PR. Otherwise I'll create a new issue for it.

File: implementations/rust/ockam/ockam_api/src/nodes/models/services.rs [hunk 10]
```diff
@@ -365,7 +365,9 @@ impl StartOktaIdentityProviderRequest {
 #[rustfmt::skip]
 #[cbor(map)]
 pub struct ServiceStatus {
+	#[serde(rename = "address", serialize_with = "crate::util::serialize_as_multiaddr")]
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/models/services.rs]: If we do this, then the `deserialize_with` attribute should be defined as well. Try to write a unit test to check that the serde methods work as expected.

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs (hunk 9 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: Yes, I completely agree. In fact, in my change to the inlets and outlets list I got rid of the list wrapper to reduce complexity and I think that it's way better. I agree that the change can be out of scope of the PR but I would gladly take on it after (or before) this one. You can create the issue and assign it to me.

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs (hunk 2 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: Fixed

File: implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs (hunk 8 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_api/src/nodes/models/portal.rs]: Fixed

File: implementations/rust/ockam/ockam_api/src/nodes/models/services.rs (hunk 10 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_api/src/nodes/models/services.rs]: Would you prefer making the same approach as we did with inlets and outlets? Creating a version of the struct to handle the custom serialization without affecting api? I can also do the deserialization and tests. Whatever you think is better.

**Comment by kriogenia** [comment]
@adrianbenavides I updated the whole PR to use custom structs instead of modifying those at `ockam_api`. Now the PR features only changes in `ockam_command`. If you want to get rid of the list wrappers as we discussed I can take into the issue if you want.

File: implementations/rust/ockam/ockam_command/src/node/models/show.rs [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+use super::{
+    portal::{ShowInletStatus, ShowOutletStatus},
+    secure_channel::ShowSecureChannelListener,
+    services::ShowServiceStatus,
+    transport::ShowTransportStatus,
+};
+
+/// Information to display in the `ockam node show` command
+#[derive(Debug, Serialize)]
+pub struct ShowNodeResponse<'a> {
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/node/models/show.rs]: I think you could get rid of that lifetime. It's only used by one of the fields, so it's not doing much. Removing it will also simplify the definition of the traits it implements.

File: implementations/rust/ockam/ockam_command/src/node/models/show.rs [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub identity: Option<String>,
+    pub transports: Vec<ShowTransportStatus>,
+    pub secure_channel_listeners: Vec<ShowSecureChannelListener>,
+    pub inlets: Vec<ShowInletStatus>,
+    pub outlets: Vec<ShowOutletStatus>,
+    pub services: Vec<ShowServiceStatus>,
+}
+#[derive(Debug, Serialize)]
+struct RouteToNode {
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/node/models/show.rs]: It fails to compile because this struct is not public. Make sure to run `cargo fmt` and `cargo clippy` to make the compiler happy 🤖

**Review by adrianbenavides — COMMENTED** [review]
Awesome! I think it feels way simpler and predictable now 👍  I left a couple of minor comments, but we can merge the PR as soon as those are fixed.

>  If you want to get rid of the list wrappers as we discussed I can take into the issue if you want.

Don't worry about that, the PR is perfect as it is now!

File: implementations/rust/ockam/ockam_api/src/nodes/models/services.rs (hunk 10 above)
Comment (adrianbenavides) [reply implementations/rust/ockam/ockam_api/src/nodes/models/services.rs]: I think you already solved this, so no custom deserialization or tests are needed.

File: implementations/rust/ockam/ockam_command/src/node/models/show.rs (hunk 12 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_command/src/node/models/show.rs]: Oh mb, I forgot that the type was being exposed directly from the properties. Fixed.

File: implementations/rust/ockam/ockam_command/src/node/models/show.rs (hunk 11 above)
Comment (kriogenia) [reply implementations/rust/ockam/ockam_command/src/node/models/show.rs]: Removed

**Review by adrianbenavides — APPROVED** [review]
Good to merge! Thank you so much @kriogenia for this awesome PR, we've been looking forward to this for quite a while now 💪
